Kwxm/specify untyped plc

kwxm opened the pull request:
I've opened this PR just for comments, ~so please don't merge it~: amongst other things, there are remarks that should be removed before merging.  There's also quite a bit about on/off chain execution that will have to be fixed, for instance stuff about on-chain typechecking.

I'll also be adding specifications of typed and untyped CEK machines at some point, possibly in another PR.

michaelpj commented:
This basically looks good!

> I am not a huge fan of having untyped values being part of the untyped term syntax. I see how it works very well here but I would like to keep the style of presentation of the untyped syntax as close as possible to that of the typed syntax. I think this makes it easier to understand how they correspond.

I agree with this.

kwxm commented:
> I am not a huge fan of having untyped values being part of the untyped term syntax. 

Yes, I hesitated about that.  I'll make it the same as the typed version.

It's always annoyed me a bit having values as part of the syntax, since then some things belong to more than one syntactic category.  I was tempted to move values to a separate figure (or subfigure) and say "these are the things that can't evaluate any further, and we can prove that".  I suppose you'd have to do the same thing for types.

kwxm commented:
> > I was tempted to move values to a separate figure (or subfigure) and say "these are the things that can't evaluate any further, and we can prove that". I suppose you'd have to do the same thing for types.
> 
> I like this - if you wanted to do this I think it would be a good idea.

OK, I'll do that.  I think we couldn't do it before because of the value restriction, but it should be OK now.

michaelpj commented:
> OK, I'll do that. I think we couldn't do it before because of the value restriction, but it should be OK now.

Yeah, that might well be the reason!

kwxm commented:
I think this should be abut ready for merging now.  I've updated the explanatory text to talk about untyped PLC and separated values and type values into separate figures in the various grammars, as discussed above.

I've added horizontal rules between some of the figures for clarity, but not consistently.  I may do a bit more fiddling with this, or use subfigures for values in the grammars instead.

Could someone (@jmchapman ?) please take a look at the appendices about algorithmic versions of things and make sure they're still consistent with the main text? 

Things still to be done
  * Add typed and untyped versions of the CEK machine (should be fairly quick)
  * Specify the extensibility mechanism for builtins  (maybe not so quick)
  * Eventually we should add a specification of the cost model and how costing is done during execution. I hope we can get away with doing that for untyped PLC only.

kwxm commented:
Oops, don't review this just yet.  I've just noticed that its still talks about types on-chain.  I'll fix that very soon.

michaelpj commented:
> I've added horizontal rules between some of the figures for clarity, but not consistently. I may do a bit more fiddling with this, or use subfigures for values in the grammars instead.

Maybe you could do what we tried to do in one of the papers, and have rules above and below figures consistently?

michaelpj commented on plutus-core-spec/plutus-core-specification.tex:
```diff
@@ -1033,6 +1027,99 @@ \section{Examples}
+%% but I think it's unlikely that it'll confuse people. Other
+%% possiblities: $\left|M\right|$, $\left\lVert{M}\right\rVert$,
+%% $\overline{M}$ (that looks pretty bad), etc. I'll consult some experts to
+%% find out if there's a standard notation.  (Later: apparently not.)
```
Seems fine. Random paper I looked at (https://ecommons.cornell.edu/bitstream/handle/1813/7375/98-1721.pdf?sequence=1) uses `M^o` or similar, which isn't very nice either.

If people complain it looks like evaluation, we can just say it *is* a kind of evaluation :D

kwxm replied:
That's my thinking too!

michaelpj commented on plutus-core-spec/plutus-core-specification.tex:
```diff
@@ -1033,6 +1027,99 @@ \section{Examples}
+for untyped Plutus Core, analogously to
+Figure~\ref{fig:term-reduction}.  The built-in functions in
+Figure~\ref{fig:builtins} are interpreted in the untyped setting in the
+obvious way.
```
I'm not sure that this is obvious! TBH, the semantics of builtins in the spec are underspecified, e.g. Fig 15 gives semantics in terms of integers for integer operations, but doesn't specify what *term* that produces. Probably we need some rule for builtins like:
- A builtin can only reduce to:
    - A `con`
    - One if its arguments (to handle polymorphic builtins)

Then I think we'd be okay with erasure.

kwxm replied:
Yeah, I fudged that a bit; I'll think about that a bit more.  We'll have to deal with extensible builtins at some point, and that may be a bit tricky. I think we'll need two explanations: a technical one saying how things work in the Haskell implementation (probably a separate document), and some more general thing in this document saying that you're free to extend the builtins however your evaluator is implemented (and no matter what language you're using), but you have to satisfy some yet-to-be-determined restrictions.  For example, presumably polymorphic builtins have to be parametrically polymorphic, to avoid having dynamically-typed Plutus Core, which would surely be a bad thing.

kwxm replied:
Having slept on this, I've decided that it's a real nuisance.  Without the types, there's nothing to stop us applying builtins to nonsensical arguments, and I don't think there's any way to overcome this without placing a lot of conditions on how the builtins behave.  I've made an attempt at dealing with this by saying that the untyped semantics are defined only for terms which we've got by erasing typed ones and then lifting the behaviour of builtins on untyped terms back to the typed semantics.  I think this means that you really do have to confirm off-chain that your untyped validator is obtained from a well-typed one (which you actually have to be in possession of, because type reconstruction is undecidable) . If you allow arbitrary untyped code to run on-chain and your evaluator happens to be implemented in C then all bets are off if a builtin is fed get arguments of the wrong sort.

michaelpj commented on plutus-core-spec/plutus-core-specification.tex:
```diff
@@ -1033,6 +1027,99 @@ \section{Examples}
+Figure~\ref{fig:untyped-ck-machine} describes a modified version of
+the CK machine (Figure~\ref{fig:ck-machine}) which evaluates untyped Plutus
+Core programs.
+\subfile{figures/UntypedCkMachine}
```
Could we do some highlighting on the figures to point out the differences?

kwxm replied:
I think there are too many differences for that. Also some bit have just disappeared, so they'd be tricky to highlight.

michaelpj replied:
Are there? Most of the rules are the same, but you've added force/delay.

I agree you can't highlight deletions, but that's okay.

michaelpj commented on plutus-core-spec/plutus-core-specification.tex:
```diff
@@ -93,7 +95,18 @@
+\newcommand{\erase}[1]{\llbracket#1\rrbracket}
+%\newcommand{\erase}[1]{\left|#1\right|}
+%\newcommand{\erase}[1]{\overline{#1}}
+%\newcommand{\erase}[1]{\left\lVert#1\right\rVert}
```
Pick one and keep it!

kwxm replied:
Thanks.  I stuck these in when I was experimenting to see which one looked best and then forgot about them.

kwxm commented on plutus-core-spec/plutus-core-specification.tex:
```diff
@@ -1033,6 +1027,99 @@ \section{Examples}
+%% immediately be replaced by $e$ alone, and then using $\lamU{x}{x}$
+%% might not cause so much overhead. I'll look into this.  Later:
+%% nope.  I looked at the on-chain validators for the use cases and
+%% this only happens about 10\% of the time.
```
Perhaps.  I was leaving it there so I could whip it out if anyone complained about force/delay.

kwxm replied:
I tried making some of it into a footnote, but that completely messed up the formatting of the rest of the document.  I've left the footnote there but commented in out in case later revision gives us space to restore it.

kwxm commented on plutus-core-spec/plutus-core-specification.tex:
```diff
@@ -1033,6 +1025,81 @@ \section{Examples}
+$P$ we use $\erase{P}$ to denote the untyped program with the same
+version whose body is the erasure of the body of $P$. The erasure operation
+preserves semantics in an appropriate sense.  It can be shown that if $P$
+is a well-typed Plutus Core program then $P$ terminates if 
```
We haven't actually shown this, but I'm confident in the abilities of our formalisers.

kwxm commented on plutus-core-spec/plutus-core-specification.tex:
```diff
@@ -1033,6 +1025,81 @@ \section{Examples}
+$\erase{M^*} = \erase{M_1} \ldots \erase{M_n}$.  For a typed program
+$P$ we use $\erase{P}$ to denote the untyped program with the same
+version whose body is the erasure of the body of $P$. The erasure operation
+preserves semantics in an appropriate sense.  It can be shown that if $P$
```
We haven't actually shown this, but I'm confident in the abilities of our formalisers.

kwxm commented on plutus-core-spec/plutus-core-specification.tex:
```diff
@@ -1033,6 +1025,81 @@ \section{Examples}
+preserves semantics in an appropriate sense.  It can be shown that if $P$
+is a well-typed Plutus Core program then $P$ terminates if 
+and only if $\erase{P}$ does, and that $P$ returns the value $v$
+if and only if $\erase{P}$ returns the value $\erase{v}$.
```
We haven't actually shown this, but I'm confident in the abilities of our formalisers.

michaelpj commented:
This conflicts because I deleted some things!

kwxm commented:
> This conflicts because I deleted some things!

I think I deleted the same things.  The conflict here is because I changed the labels on the inference rules to `textsf` in a file which we both renamed (to the same thing).  I'll just resolve the conflict here.

michaelpj commented:
Not sure if you're done with this, but feel free to merge if so.

kwxm commented:
> Not sure if you're done with this, but feel free to merge if so.

Not quite, but I'll tidy it up in the light of our earlier discussions and merge it tomorrow morning.

kwxm commented:
I'm going to merge this now.  I've added a note (7.1.1) attempting to explain how builtins are expected to behave in the untyped world, but that's necessarily a bit vague because we don't have a proper specification of builtins. We should return to this once #487 is finished.
